politeiawww/addresswatcher.go

```diff
@@ -0,0 +1,314 @@
+	if err != nil {
+		log.Errorf("failed to get server version: %v", err)
+		return err
+	}
+
+	clientSemVer := client.Version()
+	log.Infof("PubSub Server version: %s, Client version %v", serverVer, clientSemVer)
+	serverSemVer := semver.NewSemver(serverVer.Major, serverVer.Minor, serverVer.Patch)
+	if !semver.Compatible(clientSemVer, serverSemVer) {
+		return fmt.Errorf("pubsub server version is %v, but client is version %v",
+			serverSemVer, clientSemVer)
+	}
+
+	go func() {
+		for {
+			msg, ok := <-p.wsClient.Receive()
+			if !ok {
+				break
+			}
+			if msg == nil {
+				log.Errorf("ReceiveMsg failed")
+				continue
+			}
+
+			switch m := msg.Message.(type) {
+			case string:
+				log.Debugf("Message (%s): %s", msg.EventId, m)
+			case int:
+				log.Debugf("Message (%s): %v", msg.EventId, m)
+			case *pstypes.AddressMessage:
+				log.Debugf("Message (%s): AddressMessage(address=%s, txHash=%s)",
+					msg.EventId, m.Address, m.TxHash)
+
+				// Check payment history for address
+				payment, err := p.cmsDB.PaymentsByAddress(m.Address)
+				if err != nil {
+					log.Errorf("error retreiving payments information from db %v", err)
+				}
+				paid := p.checkPayments(payment)
+				if paid {
+					p.removeWatchAddress(payment.Address)
+				}
+			case *pstypes.TxList:
+				log.Debugf("Message (%s): TxList(len=%d)", msg.EventId, len(*m))
+			default:
+				log.Debugf("Message of type %v unhandled. %v", msg.EventId, m)
+			}
+		}
+	}()
+	return nil
+}
+
+func (p *politeiawww) restartAddressesWatching() error {
+	approvedInvoices, err := p.cmsDB.InvoicesByStatus(int(cms.InvoiceStatusApproved))
+	if err != nil {
+		return err
+	}
+	for _, invoice := range approvedInvoices {
+		_, err := p.cmsDB.PaymentsByAddress(invoice.PaymentAddress)
+		if err != nil {
+			if err == database.ErrInvoiceNotFound {
+				payout, err := p.calculatePayout(invoice)
+				if err != nil {
+					return err
+				}
+				// Start listening the first day of the next month of invoice.
+				listenStartDate := time.Date(int(invoice.Year),
+					time.Month(invoice.Month+1), 0, 0, 0, 0, 0, time.UTC)
+				invoice.Payments = database.Payments{
+					Address:      invoice.PaymentAddress,
+					TimeStarted:  listenStartDate.Unix(),
+					Status:       cms.PaymentStatusWatching,
+					AmountNeeded: int64(payout.DCRTotal),
+				}
+				fmt.Println(listenStartDate)
+				err = p.cmsDB.UpdateInvoice(&invoice)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	unpaidPayments, err := p.cmsDB.PaymentsByStatus(uint(cms.PaymentStatusWatching))
+	if err != nil {
+		return err
+	}
+	for _, payments := range unpaidPayments {
+		paid := p.checkPayments(&payments)
+		if !paid {
+			p.addWatchAddress(payments.Address)
+		}
+	}
+	return nil
+}
+
+func strInSlice(sl []string, str string) (bool, int) {
+	for i, s := range sl {
+		if s == str {
+			return true, i
+		}
+	}
+	return false, -1
+}
+
+func (p *politeiawww) checkPayments(payment *database.Payments) bool {
```

Comment:
This needs a comment on what it does and what the returned bool represents.  It's not obvious from just looking at it.